handle case when microsoft barfs on itself

TomzBench opened the pull request:
Fixes this issue:
https://github.com/cmake-js/cmake-js/issues/232

Solves the intent of this PR:
https://github.com/cmake-js/cmake-js/pull/225

The culprit is Microsoft - they decided they "wontfix"
https://github.com/dotnet/msbuild/issues/5726

Basically a default Windows 10 environment has a variable called Path (note the case). Some how in the node process this gets duped as PATH. Microsoft sees two path's in their environment and crashes.

davidgurnsey commented:
Up vote!

jalfd commented:
Can we be sure that we always want to keep PATH and throw away Path, and not the other way around?
I assume they can have different contents, and then deleting the wrong one might cause new problems.

TomzBench commented:
@jalfd 

On my system the two path's where completely identical. However to be sure i ended up merging the two paths like this:

```
  if (process.env.Path && process.env.PATH) {
    const env = { ...process.env };
    const PATH = [...env.Path.split(";"), ...env.Path.split(";")];
    env.PATH = [...new Set(PATH)].join(";");
    delete env.Path;
  }
  ```
  
  This leaves the real env un touched but you have to use the cloned env in your spawning. This is probably the best way to avoid unintended consequences

aminya commented:
I think this PR related to
https://github.com/cmake-js/cmake-js/pull/225

ignaszi commented:
This is an insane bug driving me mental the whole day. Could you please just merge this?

gentunian commented:
@TomzBench  Looks like removing `Path` before spawning makes `cmake-js` not able to find `cmake`:

```
ERR! OMG spawn cmake ENOENT
```

`package.json`:
```
"devDependencies": {
    "cmake-js": "github:cmake-js/cmake-js#e311b71564be3925f35f0861f85817a823d4cff0",
```

did I missed something?

gentunian commented:
Effectively, removing `process.env.PATH` instead of `process.env.Path` cmake is being called correctly in my case.

gentunian commented:
> @jalfd
> 
> On my system the two path's were completely identical. However to be sure i ended up merging the two paths like this:
> 
> ```js
>   if (process.env.Path && process.env.PATH) {
>     const env = { ...process.env };
>     const PATH = [...env.Path.split(";"), ...env.PATH.split(";")];
>     env.PATH = [...new Set(PATH)].join(";");
>     delete env.Path;
>   }
> ```
> 
> This leaves the real env un touched but you have to use the cloned env in your spawning. This is probably the best way to avoid unintended consequences

@TomzBench do you want me to create a PR with this change instead?

gentunian commented:
Ok, here I go again with a cleaner post. It looks like with this change, `cmake-js` is not able to spawn `cmake` on a second run (that's why I got too many false positives and my comments were misleading).

Consider this output:

```
13:24:13 > cmake-js compile
13:24:13 
13:24:13 [
13:24:13   '<path-hidden>\node\\node-v12.18.3-win-x64\\node.exe',
13:24:13   '<path-hidden>\node_modules\\cmake-js\\bin\\cmake-js',
13:24:13   'compile'
13:24:13 ]
13:24:14 info TOOL Using Visual Studio 16 2019 generator.
13:24:14 info CMD CONFIGURE
13:24:14 info RUN [
13:24:14 info RUN   'cmake',
13:24:14 info RUN   '<path-hidden>\',
13:24:14 info RUN   '--no-warn-unused-cli',
13:24:14 info RUN   '-G',
13:24:14 info RUN   'Visual Studio 16 2019',
13:24:14 info RUN   '-A',
13:24:14 info RUN   'x64',
13:24:14 info RUN   '-DCMAKE_JS_VERSION=6.1.0',
13:24:14 info RUN   '-DCMAKE_BUILD_TYPE=Release',
13:24:14 info RUN   '-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=<path-hidden>\',
13:24:14 info RUN   '-DCMAKE_JS_INC=<path-hidden>\',
13:24:14 info RUN   '-DCMAKE_JS_SRC=<path-hidden>',
13:24:14 info RUN   '-DNODE_RUNTIME=node',
13:24:14 info RUN   '-DNODE_RUNTIMEVERSION=12.18.3',
13:24:14 info RUN   '-DNODE_ARCH=x64',
13:24:14 info RUN   '-DCMAKE_JS_LIB=<path-hidden>\',
13:24:14 info RUN   '-DCMAKE_SHARED_LINKER_FLAGS=/DELAYLOAD:NODE.EXE'
13:24:14 info RUN ]
13:24:14 info REP Build has been failed, trying to do a full rebuild.
13:24:14 info CMD CLEAN
13:24:14 info RUN [ 'cmake', '-E', 'remove_directory', '<path-hidden>\\build' ]
13:24:14 ERR! OMG spawn cmake ENOENT
```

The problem by deleting `process.env.Path` in git-bash is still there. I will create a PR for it and I will keep you posted

gentunian commented:
@TomzBench do you think this change works? My guess is that it doesn't and somehow after deleting one we should ensure to have recreated it:

within git-bash and based on: https://nodejs.org/api/process.html#process_process_env

> On Windows operating systems, environment variables are case-insensitive.



```
$ node
Welcome to Node.js v12.17.0.
Type ".help" for more information.
> process.env.FOO = 1
1
> process.env.foo
'1'
> delete process.env.foo
true
> process.env.FOO
undefined
>
```
